src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalUtils.ts

```diff
@@ -0,0 +1,162 @@
+import _ from "lodash";
+import {fillGeneticTrackDatum} from "../../../../shared/components/oncoprint/DataUtils";
+import {OncoprinterGeneticTrackDatum, OncoprinterGeneticTrackDatum_Data} from "./OncoprinterGeneticUtils";
+import {CustomTrackOption} from "oncoprintjs";
+import {ClinicalTrackDatum, ClinicalTrackSpec} from "../../../../shared/components/oncoprint/Oncoprint";
+
+export const ONCOPRINTER_CLINICAL_VAL_NA = "N/A";
+
+export type OncoprinterClinicalInputLine = {
+    sampleId:string;
+    orderedValues:string[];
+};
+
+type AttributeSpec = { clinicalAttributeName:string, datatype:ClinicalTrackDataType };
+
+type OncoprinterClinicalTrackDatum = Pick<ClinicalTrackDatum, "attr_id"|"uid"|"attr_val_counts"|"na"> & {sample:string, attr_val:string|number};
+
+const ATTRIBUTE_REGEX = /^((?:[^\(\)])+)(?:\(([^\(\)]+)\))?$/;
+
+enum ClinicalTrackDataType {
+    NUMBER="number", LOG_NUMBER="lognumber", STRING="string"
+}
+
+function parseClinicalDataHeader(headerLine:string[]) {
+    // we dont care about the first column, it's just "sample" or something
+    headerLine.shift();
+
+    const errorPrefix = "Clinical data input error in line 1 (header): ";
+    const ret:AttributeSpec[] = [];
+    for (const attribute of headerLine) {
+        const match = attribute.match(ATTRIBUTE_REGEX);
+        if (!match) {
+            throw new Error(`${errorPrefix}misformatted attribute name ${attribute}`);
+        }
+        const datatype = (match[2] || ClinicalTrackDataType.STRING).toLowerCase();
+        if (
+            datatype !== ClinicalTrackDataType.NUMBER &&
+            datatype !== ClinicalTrackDataType.LOG_NUMBER &&
+            datatype !== ClinicalTrackDataType.STRING
+        ) {
+            throw new Error(`${errorPrefix}invalid track data type ${datatype}`);
+        }
+        ret.push({
+            clinicalAttributeName: match[1],
+            datatype: datatype as ClinicalTrackDataType
+        });
+    }
+
+    return ret;
+}
+
+export function parseClinicalInput(input:string):{ status:"complete", result:{ headers:AttributeSpec[], data:OncoprinterClinicalInputLine[]}, error:undefined }|{ status:"error", result:undefined, error:string } {
```

Comment:
maybe a comment that the response type mimics mobxpromise?

Reply:
sure, but just for the record it's not meant to be used in "await", its just a convenient form for giving back the result